File: frontend/src/services/APIService.js
// =============================================================================
// APIService.js - Centralised API Communication
// =============================================================================
// This service handles all HTTP requests to the backend API.
//
// WHY A CENTRALISED SERVICE?
// --------------------------
// Instead of calling fetch() directly in each component, we centralise all
// API calls here. This gives us:
// 1. Consistent error handling across the app
// 2. Easy to change the base URL or add authentication headers
// 3. Single place to add loading states, retries, or caching
// 4. Easier testing - we can mock this service
//
// USAGE IN COMPONENTS:
// --------------------
// import APIService from '../services/APIService';
//
// // In an async function or useEffect:
// const data = await APIService.getWeekTimetable('2025-02-03');
// =============================================================================

// Base URL for API requests
// In development, Vite's proxy forwards /api/* to the backend
// In production, this would be configured based on deployment
const API_BASE_URL = '/api';

// =============================================================================
// Helper: Make HTTP Request
// =============================================================================
// A generic fetch wrapper that handles:
// - Adding common headers
// - Parsing JSON responses
// - Error handling
// =============================================================================
async function makeRequest(endpoint, options = {}) {
  // Build the full URL
  const url = `${API_BASE_URL}${endpoint}`;

  // Default headers for JSON APIs
  const defaultHeaders = {
    'Content-Type': 'application/json',
  };

  // Merge default options with provided options
  const config = {
    ...options,
    headers: {
      ...defaultHeaders,
      ...options.headers,
    },
  };

  // If there's a body and it's an object, stringify it
  if (config.body && typeof config.body === 'object') {
    config.body = JSON.stringify(config.body);
  }

  try {
    // Make the fetch request
    const response = await fetch(url, config);

    // Handle non-OK responses (4xx, 5xx status codes)
    if (!response.ok) {
      // Try to get error details from response body
      let errorMessage;
      try {
        const errorData = await response.json();
        errorMessage = errorData.detail || errorData.message || `HTTP ${response.status}`;
      } catch {
        errorMessage = `HTTP ${response.status}: ${response.statusText}`;
      }
      throw new Error(errorMessage);
    }

    // Handle 204 No Content (e.g., successful DELETE)
    if (response.status === 204) {
      return null;
    }

    // Parse and return JSON response
    return await response.json();
  } catch (error) {
    // Log error for debugging
    console.error(`API Error [${options.method || 'GET'} ${endpoint}]:`, error);
    throw error;
  }
}

// =============================================================================
// API Service Object
// =============================================================================
// Exports methods for each API endpoint. Each method returns a Promise.
// =============================================================================
const APIService = {
  // ---------------------------------------------------------------------------
  // SETTINGS
  // ---------------------------------------------------------------------------

  /**
   * Get current application settings.
   * @returns {Promise<Object>} Settings object
   */
  getSettings: () => makeRequest('/settings'),

  /**
   * Update application settings.
   * @param {Object} settings - The settings to update
   * @returns {Promise<Object>} Updated settings
   */
  updateSettings: (settings) =>
    makeRequest('/settings', {
      method: 'PUT',
      body: settings,
    }),

  // ---------------------------------------------------------------------------
  // SUBJECTS
  // ---------------------------------------------------------------------------

  /**
   * Get all subjects, optionally filtered.
   * @param {Object} filters - Optional filters (academic_year, semester, is_active)
   * @returns {Promise<Array>} Array of subjects
   */
  getSubjects: (filters = {}) => {
    // Build query string from filters
    const params = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        params.append(key, value);
      }
    });
    const queryString = params.toString();
    return makeRequest(`/subjects${queryString ? `?${queryString}` : ''}`);
  },

  /**
   * Create a new subject.
   * @param {Object} subject - The subject data
   * @returns {Promise<Object>} Created subject
   */
  createSubject: (subject) =>
    makeRequest('/subjects', {
      method: 'POST',
      body: subject,
    }),

  /**
   * Update a subject.
   * @param {number} id - Subject ID
   * @param {Object} updates - Fields to update
   * @returns {Promise<Object>} Updated subject
   */
  updateSubject: (id, updates) =>
    makeRequest(`/subjects/${id}`, {
      method: 'PUT',
      body: updates,
    }),

  /**
   * Delete a subject.
   * @param {number} id - Subject ID
   * @returns {Promise<null>}
   */
  deleteSubject: (id) =>
    makeRequest(`/subjects/${id}`, {
      method: 'DELETE',
    }),

  // ---------------------------------------------------------------------------
  // LESSONS & TIMETABLE
  // ---------------------------------------------------------------------------

  /**
   * Get the timetable for a specific week.
   * This is the primary endpoint for the planner view.
   *
   * @param {string} startDate - The Monday of the week (YYYY-MM-DD format)
   * @returns {Promise<Object>} Week timetable with days, lessons, and Week A/B info
   *
   * @example
   * const week = await APIService.getWeekTimetable('2025-02-03');
   * console.log(week.primary_week); // "Week A" or "Week B"
   * console.log(week.days[0].lessons); // Monday's lessons
   */
  getWeekTimetable: (startDate) =>
    makeRequest(`/lessons/week?start_date=${startDate}`),

  /**
   * Get a specific lesson with all its details (notes, resources, todos).
   * @param {number} id - Lesson ID
   * @returns {Promise<Object>} Lesson with nested items
   */
  getLesson: (id) => makeRequest(`/lessons/${id}`),

  /**
   * Create a new lesson.
   * @param {Object} lesson - Lesson data (date, period, subject_id)
   * @returns {Promise<Object>} Created lesson
   */
  createLesson: (lesson) =>
    makeRequest('/lessons', {
      method: 'POST',
      body: lesson,
    }),

  /**
   * Update a lesson.
   * @param {number} id - Lesson ID
   * @param {Object} updates - Fields to update
   * @returns {Promise<Object>} Updated lesson
   */
  updateLesson: (id, updates) =>
    makeRequest(`/lessons/${id}`, {
      method: 'PUT',
      body: updates,
    }),

  /**
   * Delete a lesson.
   * @param {number} id - Lesson ID
   * @returns {Promise<null>}
   */
  deleteLesson: (id) =>
    makeRequest(`/lessons/${id}`, {
      method: 'DELETE',
    }),

  // ---------------------------------------------------------------------------
  // NOTES (attached to lessons)
  // ---------------------------------------------------------------------------

  /**
   * Add a note to a lesson.
   * @param {number} lessonId - The lesson to attach the note to
   * @param {Object} note - Note data (title, content)
   * @returns {Promise<Object>} Created note
   */
  createNote: (lessonId, note) =>
    makeRequest(`/lessons/${lessonId}/notes`, {
      method: 'POST',
      body: note,
    }),

  /**
   * Update a note.
   * @param {number} lessonId - Parent lesson ID
   * @param {number} noteId - Note ID
   * @param {Object} updates - Fields to update
   * @returns {Promise<Object>} Updated note
   */
  updateNote: (lessonId, noteId, updates) =>
    makeRequest(`/lessons/${lessonId}/notes/${noteId}`, {
      method: 'PUT',
      body: updates,
    }),

  /**
   * Delete a note.
   * @param {number} lessonId - Parent lesson ID
   * @param {number} noteId - Note ID
   * @returns {Promise<null>}
   */
  deleteNote: (lessonId, noteId) =>
    makeRequest(`/lessons/${lessonId}/notes/${noteId}`, {
      method: 'DELETE',
    }),

  // ---------------------------------------------------------------------------
  // RESOURCES (attached to lessons)
  // ---------------------------------------------------------------------------

  /**
   * Add a resource to a lesson.
   * @param {number} lessonId - The lesson to attach the resource to
   * @param {Object} resource - Resource data (title, url, etc.)
   * @returns {Promise<Object>} Created resource
   */
  createResource: (lessonId, resource) =>
    makeRequest(`/lessons/${lessonId}/resources`, {
      method: 'POST',
      body: resource,
    }),

  /**
   * Update a resource.
   * @param {number} lessonId - Parent lesson ID
   * @param {number} resourceId - Resource ID
   * @param {Object} updates - Fields to update
   * @returns {Promise<Object>} Updated resource
   */
  updateResource: (lessonId, resourceId, updates) =>
    makeRequest(`/lessons/${lessonId}/resources/${resourceId}`, {
      method: 'PUT',
      body: updates,
    }),

  /**
   * Delete a resource.
   * @param {number} lessonId - Parent lesson ID
   * @param {number} resourceId - Resource ID
   * @returns {Promise<null>}
   */
  deleteResource: (lessonId, resourceId) =>
    makeRequest(`/lessons/${lessonId}/resources/${resourceId}`, {
      method: 'DELETE',
    }),

  // ---------------------------------------------------------------------------
  // TODOS (attached to lessons)
  // ---------------------------------------------------------------------------

  /**
   * Add a todo to a lesson.
   * @param {number} lessonId - The lesson to attach the todo to
   * @param {Object} todo - Todo data (content, priority, due_date)
   * @returns {Promise<Object>} Created todo
   */
  createTodo: (lessonId, todo) =>
    makeRequest(`/lessons/${lessonId}/todos`, {
      method: 'POST',
      body: todo,
    }),

  /**
   * Update a todo.
   * @param {number} lessonId - Parent lesson ID
   * @param {number} todoId - Todo ID
   * @param {Object} updates - Fields to update
   * @returns {Promise<Object>} Updated todo
   */
  updateTodo: (lessonId, todoId, updates) =>
    makeRequest(`/lessons/${lessonId}/todos/${todoId}`, {
      method: 'PUT',
      body: updates,
    }),

  /**
   * Toggle a todo's completion status.
   * @param {number} lessonId - Parent lesson ID
   * @param {number} todoId - Todo ID
   * @returns {Promise<Object>} Updated todo
   */
  toggleTodo: (lessonId, todoId) =>
    makeRequest(`/lessons/${lessonId}/todos/${todoId}/toggle`, {
      method: 'PATCH',
    }),

  /**
   * Delete a todo.
   * @param {number} lessonId - Parent lesson ID
   * @param {number} todoId - Todo ID
   * @returns {Promise<null>}
   */
  deleteTodo: (lessonId, todoId) =>
    makeRequest(`/lessons/${lessonId}/todos/${todoId}`, {
      method: 'DELETE',
    }),
};

export default APIService;

File: ios/TeacherPlanner/TeacherPlanner/Services/APIService.swift
// =============================================================================
// APIService.swift - Network Communication with Backend
// =============================================================================
// This service handles all HTTP requests to the Teacher Planner backend API.
//
// SWIFT CONCEPTS USED:
// --------------------
// - async/await: Modern Swift concurrency for asynchronous operations
// - Generics: The request<T> function works with any Codable type
// - Result type: Could be used for explicit error handling (we use throws)
// - URLSession: Apple's networking API
//
// USAGE:
// ------
// let service = APIService.shared  // Singleton instance
// let settings = try await service.getSettings()
// =============================================================================

import Foundation

// =============================================================================
// APIError - Custom Error Types
// =============================================================================
// Define specific errors that can occur during API calls.

enum APIError: LocalizedError {
    case invalidURL
    case networkError(Error)
    case invalidResponse
    case httpError(statusCode: Int, message: String?)
    case decodingError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .invalidResponse:
            return "Invalid response from server"
        case .httpError(let statusCode, let message):
            return "HTTP \(statusCode): \(message ?? "Unknown error")"
        case .decodingError(let error):
            return "Failed to decode response: \(error.localizedDescription)"
        }
    }
}

// =============================================================================
// APIService - Main Service Class
// =============================================================================

class APIService {
    // -------------------------------------------------------------------------
    // Singleton Pattern
    // -------------------------------------------------------------------------
    // A singleton ensures only one instance exists throughout the app.
    // Access it via APIService.shared from anywhere.
    // -------------------------------------------------------------------------
    static let shared = APIService()

    // -------------------------------------------------------------------------
    // Configuration
    // -------------------------------------------------------------------------
    // The base URL of your backend server.
    // For local development, this would be your Mac's IP address.
    // For production, this would be your Unraid server's address.
    // -------------------------------------------------------------------------
    private let baseURL: String

    // JSON decoder configured for our date format
    private let decoder: JSONDecoder

    // JSON encoder for sending data
    private let encoder: JSONEncoder

    // -------------------------------------------------------------------------
    // Initializer
    // -------------------------------------------------------------------------
    private init() {
        // TODO: Make this configurable via Settings app or config file
        // For now, use localhost (works in simulator) or your server IP
        self.baseURL = "http://localhost:8000"

        // Configure JSON decoder
        self.decoder = JSONDecoder()

        // Configure JSON encoder
        self.encoder = JSONEncoder()
    }

    // -------------------------------------------------------------------------
    // Configure Base URL
    // -------------------------------------------------------------------------
    // Call this to change the server URL (e.g., from a settings screen)
    func configure(baseURL: String) -> APIService {
        // Note: In a real app, you'd want to persist this
        // For now, we return a new instance (or modify shared)
        return self
    }

    // =========================================================================
    // MARK: - Generic Request Methods
    // =========================================================================
    // These are the core methods that all API calls use.
    // They're generic over T: Decodable (response) and B: Encodable (request body).
    //
    // WHY TWO METHODS?
    // ----------------
    // Swift cannot encode "existential types" (like `Codable?` or `any Codable`).
    // If we wrote `body: Codable?`, the compiler would fail at encoder.encode(body)
    // because it doesn't know the concrete type at compile time.
    //
    // Solution: Use generics! We have:
    // 1. request<T>(endpoint:method:) - For requests WITHOUT a body (GET, DELETE)
    // 2. request<T, B>(endpoint:method:body:) - For requests WITH a body (POST, PUT)
    //
    // With generics, Swift knows the exact type at compile time and can encode it.
    //
    // ASYNC/AWAIT EXPLANATION:
    // ------------------------
    // - 'async' means this function performs asynchronous work
    // - 'throws' means it can throw errors
    // - Callers must use 'try await' to call this function
    // - The function suspends at 'await' points without blocking the thread
    // =========================================================================

    /// Request WITHOUT a body (for GET requests)
    private func request<T: Decodable>(
        endpoint: String,
        method: String = "GET"
    ) async throws -> T {
        // Build the URL
        guard let url = URL(string: "\(baseURL)\(endpoint)") else {
            throw APIError.invalidURL
        }

        // Create the request
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        // Make the network call
        // 'await' suspends here until the network call completes
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw APIError.networkError(error)
        }

        // Check HTTP status code
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        // Handle error status codes
        if httpResponse.statusCode < 200 || httpResponse.statusCode >= 300 {
            // Try to decode error message from response
            let errorMessage = try? decoder.decode([String: String].self, from: data)["detail"]
            throw APIError.httpError(
                statusCode: httpResponse.statusCode,
                message: errorMessage
            )
        }

        // Handle 204 No Content (for DELETE requests)
        if httpResponse.statusCode == 204 {
            // Return an empty instance if possible, or throw if T doesn't support it
            // For most cases, we won't call this with an expected return type for DELETE
            if let emptyResult = Optional<T>.none as? T {
                return emptyResult
            }
        }

        // Decode the response
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decodingError(error)
        }
    }

    /// Request WITH a body (for POST, PUT, PATCH requests)
    /// B: Encodable is a generic type parameter - the compiler knows the exact type
    /// at compile time, allowing proper encoding.
    private func request<T: Decodable, B: Encodable>(
        endpoint: String,
        method: String,
        body: B
    ) async throws -> T {
        // Build the URL
        guard let url = URL(string: "\(baseURL)\(endpoint)") else {
            throw APIError.invalidURL
        }

        // Create the request
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        // Encode the body - this works because B is a concrete generic type!
        request.httpBody = try encoder.encode(body)

        // Make the network call
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw APIError.networkError(error)
        }

        // Check HTTP status code
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        // Handle error status codes
        if httpResponse.statusCode < 200 || httpResponse.statusCode >= 300 {
            let errorMessage = try? decoder.decode([String: String].self, from: data)["detail"]
            throw APIError.httpError(
                statusCode: httpResponse.statusCode,
                message: errorMessage
            )
        }

        // Decode the response
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decodingError(error)
        }
    }

    // Special version for requests that don't return data (DELETE)
    private func requestNoContent(
        endpoint: String,
        method: String = "DELETE"
    ) async throws {
        guard let url = URL(string: "\(baseURL)\(endpoint)") else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let (_, response): (Data, URLResponse)
        do {
            (_, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw APIError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        if httpResponse.statusCode < 200 || httpResponse.statusCode >= 300 {
            throw APIError.httpError(statusCode: httpResponse.statusCode, message: nil)
        }
    }

    // =========================================================================
    // MARK: - Settings Endpoints
    // =========================================================================

    func getSettings() async throws -> Settings {
        return try await request(endpoint: "/settings")
    }

    func updateSettings(_ settings: [String: Any]) async throws -> Settings {
        // For partial updates, we need a flexible approach
        // This is a simplified version
        let jsonData = try JSONSerialization.data(withJSONObject: settings)
        guard let url = URL(string: "\(baseURL)/settings") else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = jsonData

        let (data, _) = try await URLSession.shared.data(for: request)
        return try decoder.decode(Settings.self, from: data)
    }

    // =========================================================================
    // MARK: - Subjects Endpoints
    // =========================================================================

    func getSubjects(isActive: Bool? = nil) async throws -> [Subject] {
        var endpoint = "/subjects"
        if let isActive = isActive {
            endpoint += "?is_active=\(isActive)"
        }
        return try await request(endpoint: endpoint)
    }

    func getSubject(id: Int) async throws -> Subject {
        return try await request(endpoint: "/subjects/\(id)")
    }

    // =========================================================================
    // MARK: - Lessons Endpoints
    // =========================================================================

    /// Get the timetable for a specific week
    /// - Parameter startDate: The Monday of the week (YYYY-MM-DD format)
    func getWeekTimetable(startDate: String) async throws -> WeekTimetable {
        return try await request(endpoint: "/lessons/week?start_date=\(startDate)")
    }

    /// Get a specific lesson with all its details
    func getLesson(id: Int) async throws -> Lesson {
        return try await request(endpoint: "/lessons/\(id)")
    }

    /// Create a new lesson
    func createLesson(_ lesson: CreateLessonRequest) async throws -> Lesson {
        return try await request(
            endpoint: "/lessons",
            method: "POST",
            body: lesson
        )
    }

    /// Update a lesson's subject assignment
    /// Note: We use UpdateLessonRequest struct instead of a dictionary because
    /// Swift's generic system needs concrete types for encoding - dictionaries
    /// don't satisfy the Encodable constraint properly in generic contexts.
    func updateLesson(id: Int, subjectId: Int) async throws -> Lesson {
        return try await request(
            endpoint: "/lessons/\(id)",
            method: "PUT",
            body: UpdateLessonRequest(subjectId: subjectId)
        )
    }

    // =========================================================================
    // MARK: - Notes Endpoints
    // =========================================================================

    func createNote(lessonId: Int, note: CreateNoteRequest) async throws -> Note {
        return try await request(
            endpoint: "/lessons/\(lessonId)/notes",
            method: "POST",
            body: note
        )
    }

    func deleteNote(lessonId: Int, noteId: Int) async throws {
        try await requestNoContent(
            endpoint: "/lessons/\(lessonId)/notes/\(noteId)"
        )
    }

    // =========================================================================
    // MARK: - Resources Endpoints
    // =========================================================================

    func createResource(lessonId: Int, resource: CreateResourceRequest) async throws -> Resource {
        return try await request(
            endpoint: "/lessons/\(lessonId)/resources",
            method: "POST",
            body: resource
        )
    }

    func deleteResource(lessonId: Int, resourceId: Int) async throws {
        try await requestNoContent(
            endpoint: "/lessons/\(lessonId)/resources/\(resourceId)"
        )
    }

    // =========================================================================
    // MARK: - Todos Endpoints
    // =========================================================================

    func createTodo(lessonId: Int, todo: CreateTodoRequest) async throws -> Todo {
        return try await request(
            endpoint: "/lessons/\(lessonId)/todos",
            method: "POST",
            body: todo
        )
    }

    /// Toggle a todo's completion status
    /// This is the key endpoint for checking off items!
    func toggleTodo(lessonId: Int, todoId: Int) async throws -> Todo {
        return try await request(
            endpoint: "/lessons/\(lessonId)/todos/\(todoId)/toggle",
            method: "PATCH"
        )
    }

    func deleteTodo(lessonId: Int, todoId: Int) async throws {
        try await requestNoContent(
            endpoint: "/lessons/\(lessonId)/todos/\(todoId)"
        )
    }
}

// =============================================================================
// NOTE: Dictionary Encodable Extension Removed
// =============================================================================
// We previously had an extension to make [String: Any] Encodable, but this
// approach has problems:
//
// 1. TYPE MISMATCH: Dictionary literals like ["key": 1] create [String: Int],
//    not [String: Any], so the extension wouldn't apply.
//
// 2. GENERIC CONSTRAINTS: Even with the extension, passing dictionaries to
//    generic functions with Encodable constraints doesn't work well because
//    Swift can't prove at compile time that the extension applies.
//
// 3. BEST PRACTICE: Using dedicated Codable structs (like UpdateLessonRequest)
//    is cleaner, type-safe, and provides better documentation of the API.
//
// All request bodies now use proper Codable structs defined in Models.swift.
// =============================================================================
